### DISCUSSION TITLE
totim not found in fp.utils.binaryfile.HeadFile.get_data()

### DISCUSSION THREAD

**Discussion by QingHe-UT** [modflowpy/flopy#1704] [General]
Hi all, 

I'm following  [MODFLOW Tutorial 2: Unconfined Transient Flow Model](https://flopy.readthedocs.io/en/3.3.6/_notebooks/tutorial02_mf.html) with modflow6 (modflow2005 is used in the tutorial). The simulations are run successfully, but when I tried to plot the results for different time step, an error occurred as "Exception: totim value (101.0) not found in file..." 

I've tested the plotting code with the results by modflow2005 and it works good. So I was wondering is there anything wrong with my code used here and is there any difference between modflow6 and 2005 when processing timestep information of the results?

**The time discretization I used for the simulations is:**
`#stress period (for time variant model)`
`nper = 3   #total stress period`
`pdata1 = (1,1,1.0)   #perlen,nstp,tsmult`
`pdata2 = (100, 100, 1.2)`
`pdata3 = (100, 100, 1.2)`
`steady = [True, False, False]   #True for steady-state sim, False for time-variant sim`

`#time discretization`
`tdis    = fp.mf6.ModflowTdis(sim,
                            time_units = 'DAYS',
                             nper = nper,
                             perioddata = (pdata1,pdata2,pdata3))`

**And the plot code I use is:**
`#plot`
`fname = os.path.join(modelws, f"{modelname}.hds")     #access the output head file`
`headobj = fp.utils.binaryfile.HeadFile(fname)`
`head = headobj.get_data(mflay = 0)`
`#headobj.list_records()`
`#times = headobj.get_times()`
`rec = headobj.get_data(totim=101.)`

**Then I checked the list_record of the simulated water head, it does contain the time step information I requested, as:**
>(1, 1, 1., 1., b'HEAD            ', 10, 10, 1)
(1, 2, 2.41493472e-07, 1.00000024, b'HEAD            ', 10, 10, 1)
(2, 2, 5.31285639e-07, 1.00000053, b'HEAD            ', 10, 10, 1)
(3, 2, 8.79036239e-07, 1.00000088, b'HEAD            ', 10, 10, 1)
(4, 2, 1.29633696e-06, 1.0000013, b'HEAD            ', 10, 10, 1)
(5, 2, 1.79709782e-06, 1.0000018, b'HEAD            ', 10, 10, 1)
(6, 2, 2.39801086e-06, 1.0000024, b'HEAD            ', 10, 10, 1)
(7, 2, 3.11910651e-06, 1.00000312, b'HEAD            ', 10, 10, 1)
(8, 2, 3.98442128e-06, 1.00000398, b'HEAD            ', 10, 10, 1)
(9, 2, 5.02279901e-06, 1.00000502, b'HEAD            ', 10, 10, 1)
(10, 2, 6.26885228e-06, 1.00000627, b'HEAD            ', 10, 10, 1)
(11, 2, 7.76411621e-06, 1.00000776, b'HEAD            ', 10, 10, 1)
(12, 2, 9.55843292e-06, 1.00000956, b'HEAD            ', 10, 10, 1)
(13, 2, 1.1711613e-05, 1.00001171, b'HEAD            ', 10, 10, 1)
(14, 2, 1.42954291e-05, 1.0000143, b'HEAD            ', 10, 10, 1)
(15, 2, 1.73960083e-05, 1.0000174, b'HEAD            ', 10, 10, 1)
(16, 2, 2.11167035e-05, 1.00002112, b'HEAD            ', 10, 10, 1)
(17, 2, 2.55815376e-05, 1.00002558, b'HEAD            ', 10, 10, 1)
(18, 2, 3.09393386e-05, 1.00003094, b'HEAD            ', 10, 10, 1)
(19, 2, 3.73686998e-05, 1.00003737, b'HEAD            ', 10, 10, 1)
(20, 2, 4.50839333e-05, 1.00004508, b'HEAD            ', 10, 10, 1)
(21, 2, 5.43422134e-05, 1.00005434, b'HEAD            ', 10, 10, 1)
(22, 2, 6.54521496e-05, 1.00006545, b'HEAD            ', 10, 10, 1)
(23, 2, 7.87840729e-05, 1.00007878, b'HEAD            ', 10, 10, 1)
(24, 2, 9.4782381e-05, 1.00009478, b'HEAD            ', 10, 10, 1)
(25, 2, 0.00011398, 1.00011398, b'HEAD            ', 10, 10, 1)
(26, 2, 0.00013702, 1.00013702, b'HEAD            ', 10, 10, 1)
(27, 2, 0.00016466, 1.00016466, b'HEAD            ', 10, 10, 1)
(28, 2, 0.00019784, 1.00019784, b'HEAD            ', 10, 10, 1)
(29, 2, 0.00023765, 1.00023765, b'HEAD            ', 10, 10, 1)
(30, 2, 0.00028542, 1.00028542, b'HEAD            ', 10, 10, 1)
(31, 2, 0.00034274, 1.00034274, b'HEAD            ', 10, 10, 1)
(32, 2, 0.00041153, 1.00041153, b'HEAD            ', 10, 10, 1)
(33, 2, 0.00049408, 1.00049408, b'HEAD            ', 10, 10, 1)
(34, 2, 0.00059314, 1.00059314, b'HEAD            ', 10, 10, 1)
(35, 2, 0.00071201, 1.00071201, b'HEAD            ', 10, 10, 1)
(36, 2, 0.00085465, 1.00085465, b'HEAD            ', 10, 10, 1)
(37, 2, 0.00102582, 1.00102582, b'HEAD            ', 10, 10, 1)
(38, 2, 0.00123122, 1.00123122, b'HEAD            ', 10, 10, 1)
(39, 2, 0.00147771, 1.00147771, b'HEAD            ', 10, 10, 1)
(40, 2, 0.00177349, 1.00177349, b'HEAD            ', 10, 10, 1)
(41, 2, 0.00212843, 1.00212843, b'HEAD            ', 10, 10, 1)
(42, 2, 0.00255436, 1.00255436, b'HEAD            ', 10, 10, 1)
(43, 2, 0.00306548, 1.00306548, b'HEAD            ', 10, 10, 1)
(44, 2, 0.00367881, 1.00367881, b'HEAD            ', 10, 10, 1)
(45, 2, 0.00441482, 1.00441482, b'HEAD            ', 10, 10, 1)
(46, 2, 0.00529802, 1.00529802, b'HEAD            ', 10, 10, 1)
(47, 2, 0.00635787, 1.00635787, b'HEAD            ', 10, 10, 1)
(48, 2, 0.00762968, 1.00762968, b'HEAD            ', 10, 10, 1)
(49, 2, 0.00915586, 1.00915586, b'HEAD            ', 10, 10, 1)
(50, 2, 0.01098727, 1.01098727, b'HEAD            ', 10, 10, 1)
(51, 2, 0.01318497, 1.01318497, b'HEAD            ', 10, 10, 1)
(52, 2, 0.01582221, 1.01582221, b'HEAD            ', 10, 10, 1)
(53, 2, 0.01898689, 1.01898689, b'HEAD            ', 10, 10, 1)
(54, 2, 0.02278451, 1.02278451, b'HEAD            ', 10, 10, 1)
(55, 2, 0.02734165, 1.02734165, b'HEAD            ', 10, 10, 1)
(56, 2, 0.03281022, 1.03281022, b'HEAD            ', 10, 10, 1)
(57, 2, 0.03937251, 1.03937251, b'HEAD            ', 10, 10, 1)
(58, 2, 0.04724725, 1.04724725, b'HEAD            ', 10, 10, 1)
(59, 2, 0.05669695, 1.05669695, b'HEAD            ', 10, 10, 1)
(60, 2, 0.06803658, 1.06803658, b'HEAD            ', 10, 10, 1)
(61, 2, 0.08164413, 1.08164413, b'HEAD            ', 10, 10, 1)
(62, 2, 0.0979732, 1.0979732, b'HEAD            ', 10, 10, 1)
(63, 2, 0.11756808, 1.11756808, b'HEAD            ', 10, 10, 1)
(64, 2, 0.14108194, 1.14108194, b'HEAD            ', 10, 10, 1)
(65, 2, 0.16929857, 1.16929857, b'HEAD            ', 10, 10, 1)
(66, 2, 0.20315853, 1.20315853, b'HEAD            ', 10, 10, 1)
(67, 2, 0.24379048, 1.24379048, b'HEAD            ', 10, 10, 1)
(68, 2, 0.29254881, 1.29254881, b'HEAD            ', 10, 10, 1)
(69, 2, 0.35105882, 1.35105882, b'HEAD            ', 10, 10, 1)
(70, 2, 0.42127082, 1.42127082, b'HEAD            ', 10, 10, 1)
(71, 2, 0.50552523, 1.50552523, b'HEAD            ', 10, 10, 1)
(72, 2, 0.60663051, 1.60663051, b'HEAD            ', 10, 10, 1)
(73, 2, 0.72795686, 1.72795686, b'HEAD            ', 10, 10, 1)
(74, 2, 0.87354847, 1.87354847, b'HEAD            ', 10, 10, 1)
(75, 2, 1.04825841, 2.04825841, b'HEAD            ', 10, 10, 1)
(76, 2, 1.25791033, 2.25791033, b'HEAD            ', 10, 10, 1)
(77, 2, 1.50949264, 2.50949264, b'HEAD            ', 10, 10, 1)
(78, 2, 1.81139141, 2.81139141, b'HEAD            ', 10, 10, 1)
(79, 2, 2.17366993, 3.17366993, b'HEAD            ', 10, 10, 1)
(80, 2, 2.60840415, 3.60840415, b'HEAD            ', 10, 10, 1)
(81, 2, 3.13008523, 4.13008523, b'HEAD            ', 10, 10, 1)
(82, 2, 3.75610251, 4.75610251, b'HEAD            ', 10, 10, 1)
(83, 2, 4.50732326, 5.50732326, b'HEAD            ', 10, 10, 1)
(84, 2, 5.40878815, 6.40878815, b'HEAD            ', 10, 10, 1)
(85, 2, 6.49054602, 7.49054602, b'HEAD            ', 10, 10, 1)
(86, 2, 7.78865547, 8.78865547, b'HEAD            ', 10, 10, 1)
(87, 2, 9.3463868, 10.3463868, b'HEAD            ', 10, 10, 1)
(88, 2, 11.21566441, 12.21566441, b'HEAD            ', 10, 10, 1)
(89, 2, 13.45879753, 14.45879753, b'HEAD            ', 10, 10, 1)
(90, 2, 16.15055728, 17.15055728, b'HEAD            ', 10, 10, 1)
(91, 2, 19.38066897, 20.38066897, b'HEAD            ', 10, 10, 1)
(92, 2, 23.25680301, 24.25680301, b'HEAD            ', 10, 10, 1)
(93, 2, 27.90816385, 28.90816385, b'HEAD            ', 10, 10, 1)
(94, 2, 33.48979686, 34.48979686, b'HEAD            ', 10, 10, 1)
(95, 2, 40.18775648, 41.18775648, b'HEAD            ', 10, 10, 1)
(96, 2, 48.22530802, 49.22530802, b'HEAD            ', 10, 10, 1)
(97, 2, 57.87036986, 58.87036986, b'HEAD            ', 10, 10, 1)
(98, 2, 69.44444408, 70.44444408, b'HEAD            ', 10, 10, 1)
(99, 2, 83.33333313, 84.33333313, b'HEAD            ', 10, 10, 1)
(100, 2, 100., 101., b'HEAD            ', 10, 10, 1)
(1, 3, 2.41493472e-07, 101.00000024, b'HEAD            ', 10, 10, 1)
(2, 3, 5.31285639e-07, 101.00000053, b'HEAD            ', 10, 10, 1)
(3, 3, 8.79036239e-07, 101.00000088, b'HEAD            ', 10, 10, 1)
(4, 3, 1.29633696e-06, 101.0000013, b'HEAD            ', 10, 10, 1)
(5, 3, 1.79709782e-06, 101.0000018, b'HEAD            ', 10, 10, 1)
(6, 3, 2.39801086e-06, 101.0000024, b'HEAD            ', 10, 10, 1)
(7, 3, 3.11910651e-06, 101.00000312, b'HEAD            ', 10, 10, 1)
(8, 3, 3.98442128e-06, 101.00000398, b'HEAD            ', 10, 10, 1)
(9, 3, 5.02279901e-06, 101.00000502, b'HEAD            ', 10, 10, 1)
(10, 3, 6.26885228e-06, 101.00000627, b'HEAD            ', 10, 10, 1)
(11, 3, 7.76411621e-06, 101.00000776, b'HEAD            ', 10, 10, 1)
(12, 3, 9.55843292e-06, 101.00000956, b'HEAD            ', 10, 10, 1)
(13, 3, 1.1711613e-05, 101.00001171, b'HEAD            ', 10, 10, 1)
(14, 3, 1.42954291e-05, 101.0000143, b'HEAD            ', 10, 10, 1)
(15, 3, 1.73960083e-05, 101.0000174, b'HEAD            ', 10, 10, 1)
(16, 3, 2.11167035e-05, 101.00002112, b'HEAD            ', 10, 10, 1)
(17, 3, 2.55815376e-05, 101.00002558, b'HEAD            ', 10, 10, 1)
(18, 3, 3.09393386e-05, 101.00003094, b'HEAD            ', 10, 10, 1)
(19, 3, 3.73686998e-05, 101.00003737, b'HEAD            ', 10, 10, 1)
(20, 3, 4.50839333e-05, 101.00004508, b'HEAD            ', 10, 10, 1)
(21, 3, 5.43422134e-05, 101.00005434, b'HEAD            ', 10, 10, 1)
(22, 3, 6.54521496e-05, 101.00006545, b'HEAD            ', 10, 10, 1)
(23, 3, 7.87840729e-05, 101.00007878, b'HEAD            ', 10, 10, 1)
(24, 3, 9.4782381e-05, 101.00009478, b'HEAD            ', 10, 10, 1)
(25, 3, 0.00011398, 101.00011398, b'HEAD            ', 10, 10, 1)
(26, 3, 0.00013702, 101.00013702, b'HEAD            ', 10, 10, 1)
(27, 3, 0.00016466, 101.00016466, b'HEAD            ', 10, 10, 1)
(28, 3, 0.00019784, 101.00019784, b'HEAD            ', 10, 10, 1)
(29, 3, 0.00023765, 101.00023765, b'HEAD            ', 10, 10, 1)
(30, 3, 0.00028542, 101.00028542, b'HEAD            ', 10, 10, 1)
(31, 3, 0.00034274, 101.00034274, b'HEAD            ', 10, 10, 1)
(32, 3, 0.00041153, 101.00041153, b'HEAD            ', 10, 10, 1)
(33, 3, 0.00049408, 101.00049408, b'HEAD            ', 10, 10, 1)
(34, 3, 0.00059314, 101.00059314, b'HEAD            ', 10, 10, 1)
(35, 3, 0.00071201, 101.00071201, b'HEAD            ', 10, 10, 1)
(36, 3, 0.00085465, 101.00085465, b'HEAD            ', 10, 10, 1)
(37, 3, 0.00102582, 101.00102582, b'HEAD            ', 10, 10, 1)
(38, 3, 0.00123122, 101.00123122, b'HEAD            ', 10, 10, 1)
(39, 3, 0.00147771, 101.00147771, b'HEAD            ', 10, 10, 1)
(40, 3, 0.00177349, 101.00177349, b'HEAD            ', 10, 10, 1)
(41, 3, 0.00212843, 101.00212843, b'HEAD            ', 10, 10, 1)
(42, 3, 0.00255436, 101.00255436, b'HEAD            ', 10, 10, 1)
(43, 3, 0.00306548, 101.00306548, b'HEAD            ', 10, 10, 1)
(44, 3, 0.00367881, 101.00367881, b'HEAD            ', 10, 10, 1)
(45, 3, 0.00441482, 101.00441482, b'HEAD            ', 10, 10, 1)
(46, 3, 0.00529802, 101.00529802, b'HEAD            ', 10, 10, 1)
(47, 3, 0.00635787, 101.00635787, b'HEAD            ', 10, 10, 1)
(48, 3, 0.00762968, 101.00762968, b'HEAD            ', 10, 10, 1)
(49, 3, 0.00915586, 101.00915586, b'HEAD            ', 10, 10, 1)
(50, 3, 0.01098727, 101.01098727, b'HEAD            ', 10, 10, 1)
(51, 3, 0.01318497, 101.01318497, b'HEAD            ', 10, 10, 1)
(52, 3, 0.01582221, 101.01582221, b'HEAD            ', 10, 10, 1)
(53, 3, 0.01898689, 101.01898689, b'HEAD            ', 10, 10, 1)
(54, 3, 0.02278451, 101.02278451, b'HEAD            ', 10, 10, 1)
(55, 3, 0.02734165, 101.02734165, b'HEAD            ', 10, 10, 1)
(56, 3, 0.03281022, 101.03281022, b'HEAD            ', 10, 10, 1)
(57, 3, 0.03937251, 101.03937251, b'HEAD            ', 10, 10, 1)
(58, 3, 0.04724725, 101.04724725, b'HEAD            ', 10, 10, 1)
(59, 3, 0.05669695, 101.05669695, b'HEAD            ', 10, 10, 1)
(60, 3, 0.06803658, 101.06803658, b'HEAD            ', 10, 10, 1)
(61, 3, 0.08164413, 101.08164413, b'HEAD            ', 10, 10, 1)
(62, 3, 0.0979732, 101.0979732, b'HEAD            ', 10, 10, 1)
(63, 3, 0.11756808, 101.11756808, b'HEAD            ', 10, 10, 1)
(64, 3, 0.14108194, 101.14108194, b'HEAD            ', 10, 10, 1)
(65, 3, 0.16929857, 101.16929857, b'HEAD            ', 10, 10, 1)
(66, 3, 0.20315853, 101.20315853, b'HEAD            ', 10, 10, 1)
(67, 3, 0.24379048, 101.24379048, b'HEAD            ', 10, 10, 1)
(68, 3, 0.29254881, 101.29254881, b'HEAD            ', 10, 10, 1)
(69, 3, 0.35105882, 101.35105882, b'HEAD            ', 10, 10, 1)
(70, 3, 0.42127082, 101.42127082, b'HEAD            ', 10, 10, 1)
(71, 3, 0.50552523, 101.50552523, b'HEAD            ', 10, 10, 1)
(72, 3, 0.60663051, 101.60663051, b'HEAD            ', 10, 10, 1)
(73, 3, 0.72795686, 101.72795686, b'HEAD            ', 10, 10, 1)
(74, 3, 0.87354847, 101.87354847, b'HEAD            ', 10, 10, 1)
(75, 3, 1.04825841, 102.04825841, b'HEAD            ', 10, 10, 1)
(76, 3, 1.25791033, 102.25791033, b'HEAD            ', 10, 10, 1)
(77, 3, 1.50949264, 102.50949264, b'HEAD            ', 10, 10, 1)
(78, 3, 1.81139141, 102.81139141, b'HEAD            ', 10, 10, 1)
(79, 3, 2.17366993, 103.17366993, b'HEAD            ', 10, 10, 1)
(80, 3, 2.60840415, 103.60840415, b'HEAD            ', 10, 10, 1)
(81, 3, 3.13008523, 104.13008523, b'HEAD            ', 10, 10, 1)
(82, 3, 3.75610251, 104.75610251, b'HEAD            ', 10, 10, 1)
(83, 3, 4.50732326, 105.50732326, b'HEAD            ', 10, 10, 1)
(84, 3, 5.40878815, 106.40878815, b'HEAD            ', 10, 10, 1)
(85, 3, 6.49054602, 107.49054602, b'HEAD            ', 10, 10, 1)
(86, 3, 7.78865547, 108.78865547, b'HEAD            ', 10, 10, 1)
(87, 3, 9.3463868, 110.3463868, b'HEAD            ', 10, 10, 1)
(88, 3, 11.21566441, 112.21566441, b'HEAD            ', 10, 10, 1)
(89, 3, 13.45879753, 114.45879753, b'HEAD            ', 10, 10, 1)
(90, 3, 16.15055728, 117.15055728, b'HEAD            ', 10, 10, 1)
(91, 3, 19.38066897, 120.38066897, b'HEAD            ', 10, 10, 1)
(92, 3, 23.25680301, 124.25680301, b'HEAD            ', 10, 10, 1)
(93, 3, 27.90816385, 128.90816385, b'HEAD            ', 10, 10, 1)
(94, 3, 33.48979686, 134.48979686, b'HEAD            ', 10, 10, 1)
(95, 3, 40.18775648, 141.18775648, b'HEAD            ', 10, 10, 1)
(96, 3, 48.22530802, 149.22530802, b'HEAD            ', 10, 10, 1)
(97, 3, 57.87036986, 158.87036986, b'HEAD            ', 10, 10, 1)
(98, 3, 69.44444408, 170.44444408, b'HEAD            ', 10, 10, 1)
(99, 3, 83.33333313, 184.33333313, b'HEAD            ', 10, 10, 1)
(100, 3, 100., 201., b'HEAD            ', 10, 10, 1)

**The error information then occurred:**
>Exception                                 Traceback (most recent call last)
Cell In[283], line 7
      5 headobj.list_records()
      6 #times = headobj.get_times()
----> 7 rec = headobj.get_data(totim=101.)
      8 #cbb = fp.utils.CellBudgetFile("./toy_tvar_mf05/modflowtest.cbc")

>File /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/flopy/utils/datafile.py:532, in LayerFile.get_data(self, kstpkper, idx, totim, mflay)
    529 else:
    530     totim1 = self.times[-1]
--> 532 data = self._get_data_array(totim1)
    533 if mflay is None:
    534     return data

>File /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/flopy/utils/datafile.py:428, in LayerFile._get_data_array(self, totim)
    426     if len(keyindices) == 0:
    427         msg = f"totim value ({totim}) not found in file..."
--> 428         raise Exception(msg)
    429 else:
    430     raise Exception("Data not found...")

>Exception: totim value (101.0) not found in file...

**Comment by mbakker7** [comment]
In MF6, you can get the heads data in a much easier way than in mf2005 and then get the stored times. For example, if your groundwater flow model object is called `gwf`, you can do
```
hds = gwf.output.head() # get handle to binary head file
times = hds.get_times()
```
You can see whether totim=101 is indeed stored. You are using a time step multiplier of 1.2, which means that the times are likely not nice round numbers. Another problem can be roundoff errors. But you can find out all those by using the `get_times` function.

**Reply by QingHe-UT** [reply]
Thank you for your reply! It does seem to the round number problem (`totim` = 101.00000024149337 rather than 101.0). Regarding this, I also want to ask how to define the step multiplier (`tsmult`)?  

In [flopy.mf6.modflow.mftdis module](https://flopy.readthedocs.io/en/3.3.6/source/flopy.mf6.modflow.mftdis.html?highlight=fp.mf6.ModflowTdis), it says the multiplier should be described as a relation between time step, total time steps, and period length, as:
<img width="195" alt="Screen Shot 2023-02-02 at 9 48 01" src="https://user-images.githubusercontent.com/115207091/216203372-a524ed12-07b9-4886-95a3-c8df6b9aabf6.png">
Is it necessary to follow exactly the equation here to obtain `tsmult` or we can also give it an arbitrary value ourselves?

P.S. There seems no such argument in time discretization module in MF2005, why do we need this new function?

**Reply by jlarsen-usgs** [reply]
@QingHe-UT 

Dataset 7 in the DIS Package for MODFLOW 2005 and related products includes the TSMULT parameter, as well as the `flopy.modflow.ModflowDis` object for MODFLOW 2005 and NWT DIS packages. In fact it's been around since at least MODFLOW 88 (probably longer, but that's the oldest documentation I have). TSMULT is an arbitrary time step multiplier that the user can set; for example if it is set to 1 the time step length will be uniform across a stress period. Setting this to a value greater than one can be useful for simulating abrupt changes in boundary conditions between two stress periods. The equation you show describes TSMULT's relationship to the first timestep within a stress period. Each successive time step length is equivalent to `delta_t_i = delta_t_i-1 * tsmult` where `i` is the current timestep.

**Reply by QingHe-UT** [reply]
Thank you for your explanation! I understand it now :)
